lang/csharp/src/apache/main/Util/Decimal.cs

```diff
@@ -0,0 +1,177 @@
+                throw new AvroTypeException("'decimal' requires a 'precision' property");
+
+            var precision = int.Parse(precisionVal);
+
+            if (precision <= 0)
+                throw new AvroTypeException("'decimal' requires a 'precision' property that is greater than zero");
+
+            var scale = GetLogicalScale(schema);
+
+            if (scale < 0 || scale > precision)
+                throw new AvroTypeException("'decimal' requires a 'scale' property that is zero or less than or equal to 'precision'");
+        }
+
+        /// <summary>
+        /// Converts a logical value to an instance of its base type.
+        /// </summary>
+        /// <param name="logicalValue">The logical value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the base type.</returns>        
+        public override object ConvertToBaseValue(object logicalValue, LogicalSchema schema)
+        {
+            var decimalValue = (decimal)logicalValue;
+            var unscaledInt = new BigInteger(GetUnscaledDecimalValue(decimalValue, out int scale));
+
+            if (scale != GetLogicalScale(schema))
+                throw new ArgumentOutOfRangeException("logicalValue", $"The decimal value has a scale of {scale} which cannot be encoded against a logical 'decimal' with a scale of {GetLogicalScale(schema)}");
+
+            if (decimalValue < 0)
+                unscaledInt = BigInteger.Negate(unscaledInt);
+
+            var buffer = unscaledInt.ToByteArray();
+
+            Array.Reverse(buffer); // Make the sequence of bytes big-endian
+
+            return Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (object)buffer
+                : (object)new GenericFixed((FixedSchema)schema.BaseSchema, buffer);
+        }
+
+        /// <summary>
+        /// Converts a base value to an instance of the logical type.
+        /// </summary>
+        /// <param name="baseValue">The base value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the logical type.</returns>
+        public override object ConvertToLogicalValue(object baseValue, LogicalSchema schema)
+        {
+            var buffer = Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (byte[])baseValue
+                : ((GenericFixed)baseValue).Value;
+
+            Array.Reverse(buffer); // Make the sequence of bytes little-endian
+
+            var unscaledInt = new BigInteger(buffer);
+            var isNegative = unscaledInt < 0;
+
+            var absUnscaledInt = BigInteger.Abs(unscaledInt);
+
+            buffer = absUnscaledInt.ToByteArray();
+
+            var paddedBuffer = new byte[12];
+
+            Buffer.BlockCopy(buffer, 0, paddedBuffer, 0, Math.Min(buffer.Length, 12));
```

Comment:
If buffer.Length > 12 we can't correctly read the avro data into a decimal - so would it be better to raise an exception in that case?

Reply:
@timjroberts  - thinking on this point some more - would it make sense to create a new type `AvroDecimal` which wraps a BigInteger and scale? This class could then include fromDecimal() and toDecimal() methods to bridge to the native decimal type for the majority of use cases, but still allow use with arbitrary size decimals as per the spec?